walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,179 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Collections;
+import java.util.HashSet;
 
+/**
+* FindMeetingQuery Class - Uses Event, Meeting Request, and Time Range objects
+* to schedule a meeting time given participants and unique schedules.
+*
+* A query is fielded in the following methodology:
+*      Collect busy times
+*      Calculate free windows
+*      Choose best window
+*      Figure which time works best
+*   
+*     If optional attendees
+*      Collect optional busy times
+*      Calculate those free windows
+*      Calculate intersection
+*/
 public final class FindMeetingQuery {
+
+  /**
+  * Function query
+  * This method searches for common free windows in the schedules of the attendees needed. 
+  * Optional attendees for the meeting are also considered and accommodated if possible.
+  * @param request The object containing the requesting event and details
+  * @param events A list of all of the events and attendees for the day
+  * @return A list of possible time frames that satisfy the request.
+  *
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    ArrayList<TimeRange> candidates = new ArrayList<>();
+    ArrayList<String> guests = new ArrayList<>(request.getAttendees());
+    ArrayList<String> optionals = new ArrayList<>(request.getOptionalAttendees());
+    long duration = request.getDuration();
+    
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    } else if (events.isEmpty() || guests.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (!optionals.isEmpty()) {
+      return meetingWithOptionals(guests, optionals, duration, events);
+    } else {
+      return communalFreeWindows(guests, duration, events);
+    }
+  }
+
+  /**
+  * Method meetingWithOptionals
+  * This method handles the case that there are optional attendees for this meeting.
+  * Accommodation of the optional attendees is all or nothing.
+  * @param mandatory The required meeting attendees.
+  * @param optionals The optional meeting attendees.
+  * @param duration The desired length of the meeting.
+  * @param events A list of all the events occuring during that day.
+  * @return A list of the time slots that attempts to accommodate all attendees.
+  */
+  private Collection<TimeRange> meetingWithOptionals
+      (Collection<String> mandatory, Collection<String> optionals, 
+      long duration, Collection<Event> events) {
+        
+    ArrayList<TimeRange> mandatoryWindows = (ArrayList<TimeRange>) communalFreeWindows(mandatory, duration, events);
+    ArrayList<TimeRange> optionalWindows = (ArrayList<TimeRange>) communalFreeWindows(optionals, duration, events);
+
+    if (mandatoryWindows.isEmpty()) {
+      return optionalWindows;
```

Comment:
When there are no available window for the mandatory guests, should this return empty list?

Reply:
Addressed: 00853d2